README.md

```diff
@@ -104,6 +104,166 @@ TODO: specify the encoding (byte-array to string) procedure
+
+In general empty values in the string representation are always disallowed unless
+explicitely noted otherwise. In case of conversion errors implementation must
+refuse to process the given string/binary value and report the error to the caller
+instead.
+
+While all codec types may be encoded using the standard variable length encoding,
+some may also allow being encoded into a specific static-length binary value
+without the extra length information if this is noted in the respective codec's
+description.
+
+All code examples are written in Python-based pseudo code and are optimized for
+legibility rather than speed. In general you should always use existing libraries
+and functions for performing the below conversions rather than rolling your own.
+
+#### `fspath`
+
+Encodes the given Unicode string using the system's local file system encoding.
+On Windows this encoding likely being UTF-16, while being UTF-8 on most other
+systems. It is up to the library to figure out the best encoding value for
+these kinds of strings.
+
+ * String → Binary: `str.encode(SYSTEM_FILESYSTEM_ENCODING)`
+ * Binary → String: `bytes.decode(SYSTEM_FILESYSTEM_ENCODING)`
+
+Protocols using the `fspath` encoding must not be shared between different hosts.
+
+#### `idna`
+
+Encodes the given Unicode representation according to IDNA-2008 ([RFC 5890](https://tools.ietf.org/html/rfc5890)) conventions using the [UTS-46](https://tools.ietf.org/html/rfc5890) input normalization and processing rules.
+
+ * String → Binary:
+    1. Normalize and validate the given input string according to [UTS-46 Section 4 (Processing)](https://www.unicode.org/reports/tr46/#Processing) and [UTS-46 Section 4.1 (Validity Criteria)](https://www.unicode.org/reports/tr46/#Validity_Criteria) with the following parameters:
+       * UseSTD3ASCIIRules = true
+       * CheckHyphens = true
+       * CheckBidi = true
+       * CheckJoiners = true
+       * Transitional_Processing = false
+    2. Convert the Unicode string to ASCII using the [UTS-46 Section 4.2 (ToASCII)](https://www.unicode.org/reports/tr46/#ToASCII) algorithm steps 2–6 with parameter *VerifyDnsLength* set to *true* and return the result.
+ * Binary → String:  
+   Convert the ASCII text string to Unicode according to the rules of [UTS-46 Section 4.3 (ToUnicode)](https://www.unicode.org/reports/tr46/#ToUnicode) using the same parameters as in step 1 of the *String → Binary* algorithm.
+
+Examples of libraries for performing the above steps include the [Python idna](https://pypi.org/project/idna/) library.
+
+#### `ip4`
+
+Encodes an IPv4 address according to the conventional [dot-decimal notation](https://en.wikipedia.org/wiki/Dot-decimal_notation) first specificed in [RFC 3986 section 3.2.2 page 20 § 2](https://tools.ietf.org/html/rfc3986#page-20).
+
+Protocols using this codec may encode it as binary value of exactly 4 bytes without
+an extra length value.
+
+ * String → Binary:
```

Comment:
Not sure if explicitly specifying an IPv4 address is necessary given the link to the spec.

Reply:
The linked spec is actually the URL spec that has an “aside” kind of note on how to *parse* IPv4 dotted-decimal addresses in an URL hostname. There is no formal spec on IPv4 textual address representation actually, so I thought reproducing the expected steps here might avoid ambiguity: For instance the POSIX `inet_addr` function would also allow `"127.0x000001"` as a valid input for `"127.0.0.1"`, so I wanted to be clear *which variations* are allowed by this spec and which are not.